### PR TITLE
Add migration for sass z-index fn

### PR DESCRIPTION
**Pull request by gwyneplaine**
WHY are these changes introduced?
Part of #7211 

WHAT is this pull request doing?
Adding a replace-sass-z-index migration script.

Running npx @shopify/polaris-migrator replace-sass-z-index "**/*.scss" will target z-index function usage and do the following: 

* If more than one argument exists we assume the code is invoking z-index to map-get on a custom sass map. In this case, we add a comment with a migration warning and code snippet replacing the `legacy-polaris-v8.z-index` usage with `map-get`
* If there's a calculation in the same declaration but the `legacy-polaris-v8.z-index` invocation is valid and corresponds to a polaris token, we add a comment with a migration warning and a code snippet replacing that invocation with a css variable invocation pointing at the corresponding polaris-token. 
* If its just a valid `legacy-polaris-v8.z-index` invocation that corresponds to an existing polaris token, we replace it with a css-variable invocation pointing at the corresponding polaris-token. 
* Otherwise we ignore the declaration. 

Things our codemod doesn't presently understand but should: 
* interpolations. `calc(#{legacy-polaris-v8.z-index} + 1)` for example is completely ignored by our code. As post-css interprets this as an invocation of a function called `#{legacy-polaris-v8.z-index`. It doesn't seem to understand interpolations at the moment. We can probably use something like style-lint's [hasInterpolation](https://github.dev/stylelint/stylelint/blob/eee9deb35190f46c5f964769d521fef5d7d1d7a8/lib/utils/__tests__/hasInterpolation.test.js#L5) to work around this, (ty @aaronccasanova), but not something we should worry about for first pass imo.

### CODE REVIEW

File: polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts
Comment (gwyneplaine) [inline polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts]: We have this in the replace-sass-spacing mod example @samrose3, and I don't know why
This isn't immediately obvious, I've commented it out for now as we **probably** don't want to unwittingly pass additional arguments to `legacy-polaris-v8.z-index` to `var`, but keen to know if there's an edge case or reason for `...node.nodes.slice(1)` here that I'm not aware of 🎉

File: polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts
Comment (gwyneplaine) [inline polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts]: We should replace this with `namespace()` and `isSassFunction` once those utilities make it into main from #7264

Comment (sam-b-rose) [reply polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts]: Yes, I'll see if we can get that merged today so it will be available for other migrations 👍

File: polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts
Comment (sam-b-rose) [reply polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts]: Great point, I forget why this was included. @aaronccasanova, do you recall what edge-case this was handling?

I did test the `replace-sass-spacing` migration without it and is working properly. **I'd be in favor of removing it (leaving it out)** since, to your point, given we probably don't want to be passing additional arguments 👍

**Review by sam-b-rose — APPROVED** [review]
This looks 💯 great catch on the unnecessary `...node.nodes` as well.

Feel free to `/snapit` and test and/or run on web 🚢 

We can circle back for function reuse and clean up after the other length PR is merged.

File: polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts
Comment (sam-b-rose) [reply polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts]: #7264 has been merged 🚢 those functions should now be ready for use

File: polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts
Comment (aaronccasanova) [inline polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts]: Note this is now exposed from `utils`

File: polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts
Comment (aaronccasanova) [inline polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts]: `utils` now exposes a `NamespaceOptions` type we can extend here e.g.
`inferface PluginOptions extends Options, NamespaceOptions {}`

File: polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts
Comment (aaronccasanova) [inline polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts]:
```suggestion
const hasMoreThanOneArgument = (node: FunctionNode) => node.nodes.length > 1;
```

Function nodes [should always have a `nodes` array](https://github.com/TrySound/postcss-value-parser/blob/11a4436e74930f4eaf9cbdb2063656bcce8ebf6a/lib/index.d.ts#L52). Did you run into an edge case?

File: polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts
Comment (aaronccasanova) [inline polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts]:
```suggestion
          const element = node.nodes[0]?.value ?? '';
```

The prior validation only guarantees there isn't more than one arg, not that there is args. We should be able to get this flagged by TypeScript ([with `noUncheckedIndexAccess` enabled](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true#code/MYewdgzgLgBGIBMCmEBcMDeMBuBDANgK5LrQBOAlmAOYwC+A2gLowC8MzAsAFA8D0fGAHUKUABYwABvACqYYGKTAA1kgQBJMMgAeagILBgKCJJhIwuAEb41PUJFh4iSNnEQoGABiYA6J8R4gA))

File: polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts
Comment (aaronccasanova) [reply polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts]: > I forget why this was included. @aaronccasanova, do you recall what edge-case this was handling?

I'm not aware of any edge cases. I simply restructured the existing behavior. Let's nix it!

File: polaris-migrator/src/migrations/replace-sass-z-index/tests/replace-sass-z-index.test.ts
Comment (aaronccasanova) [inline polaris-migrator/src/migrations/replace-sass-z-index/tests/replace-sass-z-index.test.ts]: Reminder to re-apply the `replace-sass-z-index` input/output fixtures

File: polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts
Comment (gwyneplaine) [reply polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts]: Good catch, was previous dealing with bad typing; no reason to have this check now 👍

File: polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts
Comment (gwyneplaine) [reply polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts]: Refactored the file to use these nice new utilities, ty 🎉

File: polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts
Comment (gwyneplaine) [reply polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts]: Nice and clean, love it

**Comment by gwyneplaine** [comment]
Added additional usecase of `z-index([validKey], $fixed-element-stacking-order)` 
which contains a larger subset of our z-index tokens.

**Comment by gwyneplaine** [comment]
Gonna let this rock after https://github.com/Shopify/web/pull/74921 is merged. 
@aaronccasanova will hold off on refactoring to your target array strategy, preference is to get this in and then iterate in a subsequent PR

**Comment by gwyneplaine** [comment]
Both web PRs for this migration have now been merged:
* Manual: https://github.com/Shopify/web/pull/74921
* Safe automatic: https://github.com/Shopify/web/pull/74918

Merging this in now, subsequent PR incoming to refactor to @aaronccasanova's collection strategy shortly 👍

**Pull request merged**
